File: algorithm/simulator/java/HeapSimulator.java
import java.util.*;

public class HeapSimulator {
    
    private List<Integer> heap;
    private boolean isMaxHeap;
    private int animationDelay;
    private boolean stepMode;
    private Scanner scanner;

    public HeapSimulator() {
        this.heap = new ArrayList<>();
        this.isMaxHeap = true;
        this.animationDelay = 500;
        this.stepMode = false;
        this.scanner = new Scanner(System.in);
    }

    private int parent(int i) { return (i - 1) / 2; }
    private int left(int i) { return 2 * i + 1; }
    private int right(int i) { return 2 * i + 2; }

    private boolean compare(int a, int b) {
        return isMaxHeap ? a > b : a < b;
    }

    private void clearScreen() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
    }

    private void waitForInput() {
        if (stepMode) {
            System.out.print("\n[Press Enter to continue...]");
            scanner.nextLine();
        } else {
            try {
                Thread.sleep(animationDelay);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    private void displayState(String message, Set<Integer> highlight, Set<Integer> comparing, Set<Integer> swapping) {
        clearScreen();
        System.out.println("═".repeat(70));
        System.out.println(" HEAP SIMULATOR - " + (isMaxHeap ? "MAX" : "MIN") + " HEAP ");
        System.out.println("═".repeat(70));
        System.out.println();
        
        if (message != null) {
            System.out.println("📍 " + message);
            System.out.println();
        }

        displayTree(highlight, comparing, swapping);
        System.out.println();
        displayArray(highlight, comparing, swapping);
        
        waitForInput();
    }

    private void displayTree(Set<Integer> highlight, Set<Integer> comparing, Set<Integer> swapping) {
        if (heap.isEmpty()) {
            System.out.println("  (empty heap)");
            return;
        }

        int height = (int) Math.ceil(Math.log(heap.size() + 1) / Math.log(2));
        int maxWidth = (int) Math.pow(2, height) * 4;
        
        System.out.println("─".repeat(70));
        System.out.println(" TREE VIEW ");
        System.out.println("─".repeat(70));

        Queue<Integer> queue = new LinkedList<>();
        queue.offer(0);
        int level = 0;
        
        while (!queue.isEmpty() && level < height) {
            int levelSize = queue.size();
            int spacing = maxWidth / (int) Math.pow(2, level + 1);
            StringBuilder line = new StringBuilder();
            StringBuilder connectors = new StringBuilder();
            
            line.append(" ".repeat(spacing / 2));
            connectors.append(" ".repeat(spacing / 2));
            
            for (int i = 0; i < levelSize; i++) {
                int idx = queue.poll();
                
                if (idx < heap.size()) {
                    String nodeStr;
                    if (swapping != null && swapping.contains(idx)) {
                        nodeStr = String.format("[%d]!", heap.get(idx));
                    } else if (comparing != null && comparing.contains(idx)) {
                        nodeStr = String.format("(%d)?", heap.get(idx));
                    } else if (highlight != null && highlight.contains(idx)) {
                        nodeStr = String.format("<%d>*", heap.get(idx));
                    } else {
                        nodeStr = String.format(" %d  ", heap.get(idx));
                    }
                    
                    line.append(String.format("%-4s", nodeStr));
                    
                    int leftIdx = left(idx);
                    int rightIdx = right(idx);
                    
                    if (leftIdx < heap.size()) {
                        connectors.append("/");
                    } else {
                        connectors.append(" ");
                    }
                    connectors.append("  ");
                    if (rightIdx < heap.size()) {
                        connectors.append("\\");
                    } else {
                        connectors.append(" ");
                    }
                    
                    queue.offer(leftIdx);
                    queue.offer(rightIdx);
                } else {
                    line.append("    ");
                    connectors.append("    ");
                    queue.offer(heap.size());
                    queue.offer(heap.size());
                }
                
                if (i < levelSize - 1) {
                    line.append(" ".repeat(spacing - 4));
                    connectors.append(" ".repeat(spacing - 4));
                }
            }
            
            System.out.println(line);
            if (level < height - 1) {
                System.out.println(connectors);
            }
            level++;
        }
    }

    private void displayArray(Set<Integer> highlight, Set<Integer> comparing, Set<Integer> swapping) {
        System.out.println("─".repeat(70));
        System.out.println(" ARRAY VIEW ");
        System.out.println("─".repeat(70));
        
        if (heap.isEmpty()) {
            System.out.println("  []");
            return;
        }

        StringBuilder indices = new StringBuilder("Index: ");
        StringBuilder values = new StringBuilder("Value: ");
        StringBuilder markers = new StringBuilder("       ");
        
        for (int i = 0; i < heap.size(); i++) {
            indices.append(String.format("[%2d] ", i));
            values.append(String.format(" %2d  ", heap.get(i)));
            
            if (swapping != null && swapping.contains(i)) {
                markers.append("  !  ");
            } else if (comparing != null && comparing.contains(i)) {
                markers.append("  ?  ");
            } else if (highlight != null && highlight.contains(i)) {
                markers.append("  *  ");
            } else {
                markers.append("     ");
            }
        }
        
        System.out.println(indices);
        System.out.println(values);
        System.out.println(markers);
        System.out.println();
        System.out.println("Legend: * = Highlight, ? = Comparing, ! = Swapping");
    }

    private void swap(int i, int j) {
        int temp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, temp);
    }

    public void insert(int value) {
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" INSERT(" + value + ") ");
        System.out.println("═".repeat(70));

        heap.add(value);
        int idx = heap.size() - 1;
        
        Set<Integer> highlight = new HashSet<>(Collections.singletonList(idx));
        displayState("Inserted " + value + " at index " + idx, highlight, null, null);

        while (idx > 0 && compare(heap.get(idx), heap.get(parent(idx)))) {
            int parentIdx = parent(idx);
            
            Set<Integer> comparing = new HashSet<>(Arrays.asList(idx, parentIdx));
            displayState(String.format("%d > parent %d at index %d", 
                        heap.get(idx), heap.get(parentIdx), parentIdx), null, comparing, null);
            
            Set<Integer> swapping = new HashSet<>(Arrays.asList(idx, parentIdx));
            displayState(String.format("Swap arr[%d]=%d <-> arr[%d]=%d", 
                        idx, heap.get(idx), parentIdx, heap.get(parentIdx)), null, null, swapping);
            
            swap(idx, parentIdx);
            idx = parentIdx;
        }
        
        highlight = new HashSet<>(Collections.singletonList(idx));
        displayState("Insert complete! " + value + " is now at index " + idx, highlight, null, null);
    }

    public Integer extract() {
        if (heap.isEmpty()) {
            System.out.println("Heap is empty!");
            return null;
        }

        System.out.println("\n" + "═".repeat(70));
        System.out.println(" EXTRACT " + (isMaxHeap ? "MAX" : "MIN") + " ");
        System.out.println("═".repeat(70));

        int extracted = heap.get(0);
        Set<Integer> highlight = new HashSet<>(Collections.singletonList(0));
        displayState("Extracting root: " + extracted, highlight, null, null);

        if (heap.size() == 1) {
            heap.remove(0);
            displayState("Heap is now empty", null, null, null);
            return extracted;
        }

        heap.set(0, heap.remove(heap.size() - 1));
        displayState("Moved last element " + heap.get(0) + " to root", highlight, null, null);

        heapifyDown(0);

        displayState("Extract complete! Removed " + extracted, null, null, null);
        return extracted;
    }

    private void heapifyDown(int idx) {
        int n = heap.size();

        while (true) {
            int target = idx;
            int leftIdx = left(idx);
            int rightIdx = right(idx);

            if (leftIdx < n && compare(heap.get(leftIdx), heap.get(target))) {
                target = leftIdx;
            }
            if (rightIdx < n && compare(heap.get(rightIdx), heap.get(target))) {
                target = rightIdx;
            }

            if (target == idx) break;

            Set<Integer> comparing = new HashSet<>(Arrays.asList(idx, target));
            displayState(String.format("Heapify: Compare arr[%d]=%d with arr[%d]=%d", 
                        idx, heap.get(idx), target, heap.get(target)), null, comparing, null);

            Set<Integer> swapping = new HashSet<>(Arrays.asList(idx, target));
            displayState(String.format("Heapify: Swap arr[%d]=%d <-> arr[%d]=%d", 
                        idx, heap.get(idx), target, heap.get(target)), null, null, swapping);
            
            swap(idx, target);
            idx = target;
        }
    }

    public void buildHeap(int[] arr) {
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" BUILD HEAP ");
        System.out.println("═".repeat(70));

        heap.clear();
        for (int val : arr) {
            heap.add(val);
        }
        
        displayState("Initial array loaded: " + heap, null, null, null);

        for (int i = heap.size() / 2 - 1; i >= 0; i--) {
            Set<Integer> highlight = new HashSet<>(Collections.singletonList(i));
            displayState("Heapifying from index " + i, highlight, null, null);
            heapifyDown(i);
        }

        displayState("Build complete! Heap: " + heap, null, null, null);
    }

    public Integer peek() {
        if (heap.isEmpty()) {
            System.out.println("Heap is empty!");
            return null;
        }
        Set<Integer> highlight = new HashSet<>(Collections.singletonList(0));
        displayState("Top element: " + heap.get(0), highlight, null, null);
        return heap.get(0);
    }

    public void clear() {
        heap.clear();
    }

    public void setHeapType(boolean maxHeap) {
        this.isMaxHeap = maxHeap;
        if (!heap.isEmpty()) {
            int[] arr = heap.stream().mapToInt(Integer::intValue).toArray();
            buildHeap(arr);
        }
    }

    public void runInteractiveMode() {
        System.out.println("═".repeat(70));
        System.out.println(" HEAP / PRIORITY QUEUE SIMULATOR ");
        System.out.println("═".repeat(70));
        System.out.println();
        System.out.println("Commands:");
        System.out.println("  insert <value>  - Insert a value");
        System.out.println("  extract         - Extract max/min");
        System.out.println("  peek            - View top element");
        System.out.println("  build <values>  - Build heap from values");
        System.out.println("  type <max/min>  - Switch heap type");
        System.out.println("  display         - Show heap state");
        System.out.println("  speed <ms>      - Set animation delay");
        System.out.println("  step            - Toggle step mode");
        System.out.println("  demo            - Run demo");
        System.out.println("  clear           - Clear heap");
        System.out.println("  quit            - Exit");
        System.out.println();

        while (true) {
            System.out.print("heap> ");
            String input = scanner.nextLine().trim().toLowerCase();
            String[] parts = input.split("\\s+");

            if (parts.length == 0 || parts[0].isEmpty()) continue;

            String cmd = parts[0];

            try {
                switch (cmd) {
                    case "quit":
                    case "exit":
                        System.out.println("Goodbye!");
                        return;

                    case "insert":
                        if (parts.length > 1) {
                            insert(Integer.parseInt(parts[1]));
                        } else {
                            System.out.println("Usage: insert <value>");
                        }
                        break;

                    case "extract":
                        Integer extracted = extract();
                        if (extracted != null) {
                            System.out.println("Extracted: " + extracted);
                        }
                        break;

                    case "peek":
                        Integer top = peek();
                        if (top != null) {
                            System.out.println("Top: " + top);
                        }
                        break;

                    case "build":
                        if (parts.length > 1) {
                            String[] values = input.substring(6).split("[,\\s]+");
                            int[] arr = Arrays.stream(values)
                                    .filter(s -> !s.isEmpty())
                                    .mapToInt(Integer::parseInt)
                                    .toArray();
                            buildHeap(arr);
                        } else {
                            System.out.println("Usage: build <values>");
                        }
                        break;

                    case "type":
                        if (parts.length > 1) {
                            setHeapType(parts[1].equals("max"));
                            System.out.println("Switched to " + (isMaxHeap ? "Max" : "Min") + " Heap");
                        }
                        break;

                    case "display":
                        displayState("Current heap state", null, null, null);
                        break;

                    case "speed":
                        if (parts.length > 1) {
                            animationDelay = Integer.parseInt(parts[1]);
                            System.out.println("Delay set to " + animationDelay + "ms");
                        }
                        break;

                    case "step":
                        stepMode = !stepMode;
                        System.out.println("Step mode: " + (stepMode ? "ON" : "OFF"));
                        break;

                    case "demo":
                        runDemo();
                        break;

                    case "clear":
                        clear();
                        System.out.println("Heap cleared.");
                        break;

                    default:
                        System.out.println("Unknown command: " + cmd);
                }
            } catch (NumberFormatException e) {
                System.out.println("Error: Please enter valid integers");
            }
        }
    }

    public void runDemo() {
        clear();
        System.out.println("\n" + "═".repeat(70));
        System.out.println(" DEMO - Heap Operations ");
        System.out.println("═".repeat(70));
        System.out.println();
        System.out.println("Demo will:");
        System.out.println("1. Insert: 30, 20, 40, 10, 50, 25, 35");
        System.out.println("2. Extract top 3 elements");
        System.out.println();
        System.out.print("Press Enter to start...");
        scanner.nextLine();

        int[] values = {30, 20, 40, 10, 50, 25, 35};
        for (int v : values) {
            insert(v);
        }

        System.out.println("\nExtracting top 3 elements...");
        System.out.print("Press Enter to continue...");
        scanner.nextLine();

        for (int i = 0; i < 3; i++) {
            Integer val = extract();
            System.out.println("Extracted: " + val);
        }

        System.out.println("\n" + "═".repeat(70));
        System.out.println(" DEMO COMPLETED ");
        System.out.println("═".repeat(70));
    }

    public static void main(String[] args) {
        HeapSimulator sim = new HeapSimulator();

        if (args.length > 0 && args[0].equals("--demo")) {
            sim.runDemo();
        } else {
            sim.runInteractiveMode();
        }
    }
}

File: code/simulator/heap/python/heap_simulator.py
#!/usr/bin/env python3
"""
Heap / Priority Queue Interactive Simulator
Max Heap, Min Heap 연산을 시각화합니다.
"""

import sys
import os
import time


class HeapSimulator:
    def __init__(self, heap_type="max"):
        self.heap = []
        self.heap_type = heap_type
        self.step_mode = False
        self.animation_delay = 0.3
        self.comparisons = 0
        self.swaps = 0

    def clear_screen(self):
        os.system("clear" if os.name != "nt" else "cls")

    def compare(self, a, b):
        self.comparisons += 1
        if self.heap_type == "max":
            return a > b
        else:
            return a < b

    def parent(self, i):
        return (i - 1) // 2

    def left_child(self, i):
        return 2 * i + 1

    def right_child(self, i):
        return 2 * i + 2

    def render_heap_array(self, highlight_indices=None, label=""):
        highlight_indices = highlight_indices or []

        print("\n" + "─" * 70)
        if label:
            print(f" {label} ".center(70))
        print("─" * 70)

        print("\n  Array representation:")
        print("  Index: ", end="")
        for i in range(len(self.heap)):
            marker = "→" if i in highlight_indices else " "
            print(f"{marker}[{i}]", end=" ")
        print()

        print("  Value: ", end="")
        for i, val in enumerate(self.heap):
            marker = "→" if i in highlight_indices else " "
            print(f"{marker}{val:3d}", end=" ")
        print()

    def render_heap_tree(self, highlight_indices=None):
        if not self.heap:
            print("  (empty heap)")
            return

        highlight_indices = highlight_indices or []

        print("\n  Tree representation:")

        n = len(self.heap)
        level = 0
        idx = 0
        max_level = 0
        temp = n
        while temp > 0:
            max_level += 1
            temp = (temp - 1) // 2

        while idx < n:
            level_size = 2**level
            indent = " " * (2 ** (max_level - level + 1))
            spacing = " " * (2 ** (max_level - level + 2))

            values = []
            for _ in range(level_size):
                if idx < n:
                    marker = "→" if idx in highlight_indices else " "
                    values.append(f"{marker}{self.heap[idx]:3d}")
                    idx += 1
                else:
                    break

            print(indent + spacing.join(values))
            level += 1

        print("\n" + "─" * 70)
        print(
            f"  {self.heap_type.upper()} HEAP | Size: {len(self.heap)} | Comparisons: {self.comparisons} | Swaps: {self.swaps}"
        )

    def render_state(self, highlight_indices=None, label=""):
        highlight_indices = highlight_indices or []

        print("=" * 70)
        print(f" {self.heap_type.upper()} HEAP STATE ".center(70))
        print("=" * 70)

        self.render_heap_array(highlight_indices, label)
        self.render_heap_tree(highlight_indices)

    def insert(self, value):
        self.clear_screen()
        print("=" * 70)
        print(f" INSERT({value}) - Starting ".center(70))
        print("=" * 70)

        self.heap.append(value)
        idx = len(self.heap) - 1

        print(f"\n  Added {value} at index {idx}")
        self.render_state(highlight_indices=[idx], label="After Insertion")

        if self.step_mode:
            input("\nPress Enter to start bubble-up...")
        else:
            time.sleep(self.animation_delay)

        while idx > 0:
            parent_idx = self.parent(idx)

            self.clear_screen()
            print("=" * 70)
            print(
                f" BUBBLE-UP: Comparing index {idx} with parent {parent_idx} ".center(
                    70
                )
            )
            print("=" * 70)

            print(f"\n  heap[{idx}] = {self.heap[idx]}")
            print(f"  heap[{parent_idx}] = {self.heap[parent_idx]}")

            self.render_state(highlight_indices=[idx, parent_idx], label="Comparing...")

            if self.compare(self.heap[idx], self.heap[parent_idx]):
                self.heap[idx], self.heap[parent_idx] = (
                    self.heap[parent_idx],
                    self.heap[idx],
                )
                self.swaps += 1

                print(f"\n  ✓ Swapped! {self.heap[parent_idx]} ↔ {self.heap[idx]}")

                if self.step_mode:
                    input("\nPress Enter to continue...")
                else:
                    time.sleep(self.animation_delay)

                idx = parent_idx
            else:
                print(f"\n  ✓ Heap property satisfied. No swap needed.")
                break

            if self.step_mode:
                input("\nPress Enter to continue...")
            else:
                time.sleep(self.animation_delay)

        self.clear_screen()
        print("=" * 70)
        print(f" INSERT({value}) - Complete! ".center(70))
        print("=" * 70)
        self.render_state(label="Final State")

    def extract(self):
        if not self.heap:
            print("❌ Heap is empty!")
            return None

        self.clear_screen()
        print("=" * 70)
        print(
            f" EXTRACT {'MAX' if self.heap_type == 'max' else 'MIN'} - Starting ".center(
                70
            )
        )
        print("=" * 70)

        root = self.heap[0]
        print(f"\n  Extracting root: {root}")

        self.render_state(highlight_indices=[0], label="Root to Extract")

        if self.step_mode:
            input("\nPress Enter to continue...")
        else:
            time.sleep(self.animation_delay)

        last = self.heap.pop()

        if not self.heap:
            print(f"\n✅ Extracted {root}. Heap is now empty.")
            return root

        self.heap[0] = last
        print(f"\n  Moved last element ({last}) to root")

        self.render_state(highlight_indices=[0], label="After Moving Last to Root")

        if self.step_mode:
            input("\nPress Enter to start bubble-down...")
        else:
            time.sleep(self.animation_delay)

        self._heapify_down(0)

        self.clear_screen()
        print("=" * 70)
        print(f" EXTRACT - Complete! Returned {root} ".center(70))
        print("=" * 70)
        self.render_state(label="Final State")

        return root

    def _heapify_down(self, idx):
        while True:
            target = idx
            left = self.left_child(idx)
            right = self.right_child(idx)

            if left < len(self.heap) and self.compare(
                self.heap[left], self.heap[target]
            ):
                target = left

            if right < len(self.heap) and self.compare(
                self.heap[right], self.heap[target]
            ):
                target = right

            if target == idx:
                break

            self.clear_screen()
            print("=" * 70)
            print(f" BUBBLE-DOWN: index {idx} → {target} ".center(70))
            print("=" * 70)

            print(f"\n  Comparing with children:")
            if left < len(self.heap):
                print(f"    Left child [{left}]: {self.heap[left]}")
            if right < len(self.heap):
                print(f"    Right child [{right}]: {self.heap[right]}")
            print(
                f"  → {'Swap' if self.heap_type == 'max' else 'Swap'} with index {target}"
            )

            self.heap[idx], self.heap[target] = self.heap[target], self.heap[idx]
            self.swaps += 1

            self.render_state(highlight_indices=[idx, target], label="After Swap")

            if self.step_mode:
                input("\nPress Enter to continue...")
            else:
                time.sleep(self.animation_delay)

            idx = target

    def peek(self):
        if not self.heap:
            print("❌ Heap is empty!")
            return None
        return self.heap[0]

    def build_heap(self, array):
        self.heap = array.copy()
        self.comparisons = 0
        self.swaps = 0

        self.clear_screen()
        print("=" * 70)
        print(" BUILD HEAP - Starting ".center(70))
        print("=" * 70)

        print(f"\n  Input array: {array}")
        self.render_state(label="Initial (Not a Heap)")

        if self.step_mode:
            input("\nPress Enter to start heapify...")
        else:
            time.sleep(self.animation_delay)

        for i in range(len(self.heap) // 2 - 1, -1, -1):
            self.clear_screen()
            print("=" * 70)
            print(f" HEAPIFY from index {i} ".center(70))
            print("=" * 70)

            self.render_state(highlight_indices=[i], label=f"Heapifying subtree at {i}")

            if self.step_mode:
                input("\nPress Enter to continue...")
            else:
                time.sleep(self.animation_delay)

            self._heapify_down_quiet(i)

        self.clear_screen()
        print("=" * 70)
        print(" BUILD HEAP - Complete! ".center(70))
        print("=" * 70)
        self.render_state(label="Valid Heap")

    def _heapify_down_quiet(self, idx):
        while True:
            target = idx
            left = self.left_child(idx)
            right = self.right_child(idx)

            if left < len(self.heap) and self.compare(
                self.heap[left], self.heap[target]
            ):
                target = left

            if right < len(self.heap) and self.compare(
                self.heap[right], self.heap[target]
            ):
                target = right

            if target == idx:
                break

            self.heap[idx], self.heap[target] = self.heap[target], self.heap[idx]
            self.swaps += 1
            idx = target

    def heap_sort(self):
        if not self.heap:
            print("❌ Heap is empty!")
            return []

        original = self.heap.copy()
        sorted_array = []

        self.clear_screen()
        print("=" * 70)
        print(" HEAP SORT - Starting ".center(70))
        print("=" * 70)

        self.render_state(label="Initial Heap")

        if self.step_mode:
            input("\nPress Enter to start...")
        else:
            time.sleep(self.animation_delay)

        while self.heap:
            val = self.extract()
            sorted_array.append(val)

            if self.step_mode:
                input(f"\nExtracted {val}. Press Enter to continue...")
            else:
                time.sleep(self.animation_delay * 0.5)

        self.heap = original

        self.clear_screen()
        print("=" * 70)
        print(" HEAP SORT - Complete! ".center(70))
        print("=" * 70)

        if self.heap_type == "max":
            print(f"\n  Sorted (descending): {sorted_array}")
            print(f"  Sorted (ascending): {sorted_array[::-1]}")
        else:
            print(f"\n  Sorted (ascending): {sorted_array}")

        return sorted_array


def print_help():
    print("""
╔══════════════════════════════════════════════════════════════════════╗
║                     HEAP SIMULATOR - HELP                            ║
╠══════════════════════════════════════════════════════════════════════╣
║ Commands:                                                            ║
║   insert <value>  - Insert a value into heap                         ║
║   extract         - Extract max/min from heap                        ║
║   peek            - View top element without removing                 ║
║   build <values>  - Build heap from array (e.g., build 5 3 8 1 9)    ║
║   sort            - Perform heap sort                                 ║
║   display         - Show current heap                                 ║
║   type <max/min>  - Switch between max/min heap                       ║
║   clear           - Clear the heap                                    ║
║   speed <mode>    - Set speed (instant/fast/normal/slow)             ║
║   step            - Toggle step-by-step mode                          ║
║   demo            - Run demo                                          ║
║   help            - Show this help                                    ║
║   quit            - Exit simulator                                    ║
╚══════════════════════════════════════════════════════════════════════╝
    """)


def run_demo_mode():
    print("\n" + "=" * 70)
    print(" DEMO MODE - Heap / Priority Queue ".center(70))
    print("=" * 70)

    print("""
This demo will show:
1. Build a max heap from array [3, 1, 4, 1, 5, 9, 2, 6]
2. Insert a new value (8)
3. Extract max values
    """)

    input("Press Enter to start...")

    sim = HeapSimulator("max")
    sim.animation_delay = 0.4

    sim.build_heap([3, 1, 4, 1, 5, 9, 2, 6])

    input("\nPress Enter to insert 8...")
    sim.insert(8)

    input("\nPress Enter to extract max...")
    val = sim.extract()
    print(f"\n  Extracted: {val}")

    input("\nPress Enter to extract another max...")
    val = sim.extract()
    print(f"\n  Extracted: {val}")

    print("\n" + "=" * 70)
    print(" DEMO COMPLETED ".center(70))
    print("=" * 70)


def run_interactive_mode():
    print("=" * 70)
    print(" HEAP / PRIORITY QUEUE SIMULATOR ".center(70))
    print("=" * 70)

    sim = HeapSimulator("max")
    print_help()

    while True:
        try:
            prompt = f"heap({sim.heap_type})> "
            cmd = input(f"\n{prompt}").strip().lower()
        except EOFError:
            break
        except KeyboardInterrupt:
            print("\n")
            break

        if not cmd:
            continue

        parts = cmd.split()
        command = parts[0]

        if command == "quit" or command == "exit":
            print("Goodbye!")
            break

        elif command == "help":
            print_help()

        elif command == "insert":
            if len(parts) < 2:
                print("Usage: insert <value>")
                continue
            try:
                value = int(parts[1])
                sim.insert(value)
            except ValueError:
                print("Error: Please enter a valid integer")

        elif command == "extract":
            sim.extract()

        elif command == "peek":
            val = sim.peek()
            if val is not None:
                print(f"Top element: {val}")

        elif command == "build":
            if len(parts) < 2:
                print("Usage: build <value1> <value2> ...")
                continue
            try:
                arr = [int(x) for x in parts[1:]]
                sim.build_heap(arr)
            except ValueError:
                print("Error: Please enter valid integers")

        elif command == "sort":
            sim.heap_sort()

        elif command == "display":
            sim.render_state(label="Current State")

        elif command == "type":
            if len(parts) < 2:
                print("Usage: type <max/min>")
                continue
            heap_type = parts[1]
            if heap_type in ["max", "min"]:
                sim.heap_type = heap_type
                sim.heap = []
                print(f"✅ Switched to {heap_type} heap (heap cleared)")
            else:
                print("Unknown type. Use: max, min")

        elif command == "clear":
            sim.heap = []
            sim.comparisons = 0
            sim.swaps = 0
            print("✅ Heap cleared")

        elif command == "speed":
            if len(parts) < 2:
                print("Usage: speed <instant/fast/normal/slow>")
                continue
            mode = parts[1]
            speeds = {"instant": 0, "fast": 0.1, "normal": 0.3, "slow": 0.8}
            if mode in speeds:
                sim.animation_delay = speeds[mode]
                print(f"✅ Speed set to {mode}")
            else:
                print("Unknown speed mode")

        elif command == "step":
            sim.step_mode = not sim.step_mode
            print(f"✅ Step-by-step mode: {'ON' if sim.step_mode else 'OFF'}")

        elif command == "demo":
            run_demo_mode()

        else:
            print(f"Unknown command: {command}. Type 'help' for available commands.")


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "--demo":
        run_demo_mode()
    else:
        run_interactive_mode()
